dartagnan/src/main/java/com/dat3m/dartagnan/program/memory/Location.java

```diff
@@ -2,18 +2,17 @@
 
 import com.dat3m.dartagnan.expression.ExpressionKind;
 import com.dat3m.dartagnan.expression.ExpressionVisitor;
+import com.dat3m.dartagnan.expression.Type;
 import com.dat3m.dartagnan.expression.base.LeafExpressionBase;
-import com.dat3m.dartagnan.expression.type.IntegerType;
-import com.dat3m.dartagnan.expression.type.TypeFactory;
 
-public class Location extends LeafExpressionBase<IntegerType> {
+public class Location extends LeafExpressionBase<Type> {
 
     private final String name;
     private final MemoryObject base;
     private final int offset;
 
-    public Location(String name, MemoryObject base, int offset) {
-        super(TypeFactory.getInstance().getArchType());
+    public Location(String name, Type type, MemoryObject base, int offset) {
+        super(type);
```

Comment:
In general, assertions can be done on types that are different than arch type, which is the case of spirv. For example, we may need assertions on 32-bit integers. That's why arch type of the type of the memory object (always arch type) does not work correctly.

Location class is used only for assertions, and I think it is not used in the right (or elegant) way. In my opinion, it should be replaced by a new (non-scoped) Pointer class, which knows address (memory object + offset) and the type used for comparison.

Adding type to Location is a hack that enabled assertions for non-64-bit values without re-writing much code that has nothing to do with spirv. But I agree that it is ugly and the whole Location should be refactored later.

Reply:
> Location class is used only for assertions, and I think it is not used in the right (or elegant) way

I agree that `Location` is a hacky class that I believe existed from way back and I only got updated to "keep working" somehow.

> Adding type to Location is a hack that enabled assertions for non-64-bit values without re-writing much code that has nothing to do with spirv.

I understand that, but you can still encode the memory content at a location using the type the initial value of that location holds, assuming that memory is well-initialized in SPIRV litmus code (and there are no mixed-type accesses). This would be "one less hack" as you only require to touch the encoding of the final location value (again, assuming the type can be inferred from the initial value).
Maybe there is a reason to make the comparison with a type different to the actual content type? Even in this case though, the encoder can actually encode comparisons between different types by extending (or truncating) one of the arguments (though this also suffers from some issues).

Reply:
I understood what you mean, but it will not work that way (without further changes). Assume you want to create an assertion comparing a location of type 32-bit integer with a constant (also 32 bit). In this case, creating a compare expression will fail because of different types: 64-bit location (the memory object type) vs a 32-bit integer.

In principle, we could convert all assertions to 64-bit, but it will not work for size greater that 64 bit, signed integers etc. Adding type to location is the cleanest fast hack that I see.

Reply:
Ah, I see what you mean. Even if we encoded the correct type for `Location`, you couldn't create the compare expression to begin with due to the type mismatch. That is indeed true. You could probably still override `getType()` to return the memory content type(*) (via `getInitialValue`) rather than specifically passing a type to the constructor, so you can avoid a possible mismatch between actual content type and declared content type.
Anyway, I misunderstood the issue at first, so I don't mind doing it the way you do.
But then can you please add a comment that says `Location` represents the memory content rather than the memory address? Cause that is indeed how we use it for Litmus code right now, subtly different to what `MemoryObject` represents.

(*) But then `Location.getType()` would only be well-defined if there is a default value already associated and I'm not sure if this is the case during parsing time of the assertions.

Reply:
The types must be known on the time of parsing assertions